nixos/nextcloud: Move options into now freeForm extraOptions

onny opened the pull request:
###### Description of changes

Changing extraOptions into freeForm type and move some existing options into it by respecting upstream naming conventions

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius commented:
Hey, thank you for the PR, how can I help you to get this to the finish line?

onny commented:
I still would like to clean up the module since in my opinion there are a lot of options which we don't need anymore. Not sure yet which we could drop and how to migrate some of them correctly. I might look into this again

SuperSandro2000 commented:
Shouldn't we make ``options`` a freeform setting instead and move all the options into there? Right now we are loosing some defaults and a lot of descriptions.

onny commented:
We could merge some defaults into extraOptions which could also be overwritten?
Instead of all the description we should reference the upstream documentation. The nextcloud module is already quite big and the cleanup would be nice.
Isn‘t the extraOptions free form already?

Ma27 commented:
Considering that it's freeform: can't _move_ those options to `extraOptions`? This has the following benefits:

* defaults are automatically documented in our documentation for free
* we save our descriptions (if the descriptions are actually useful, then `man configuration.nix` >> googling for upstream docs)
* we still have useful type checking at eval-time

while we shouldn't add every single new option to the module, I don't see why we should throw away what we have, especially since these are rather commonly used options.

Also, I'm wondering if we could move - almost - everything from `config` to `extraOptions`? That way we'd also get rid of all the ugly PHP generation stuff.

onny commented:
Thank you @Ma27 and @SuperSandro2000 . I moved the existing options into extraOptions and converted it into a freeForm type. One thing missing and I'm not sure how to integrate it: Merge `cfg.hostName` into `extraOptions.trusted_domains`

```
'trusted_domains' => ${writePhpArray ([ cfg.hostName ] ++ c.extraTrustedDomains)},
```

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -926,7 +917,7 @@ in {
               ${occ}/bin/nextcloud-occ config:system:set trusted_domains \
                 ${toString i} --value="${toString v}"
-            '') ([ cfg.hostName ] ++ cfg.config.extraTrustedDomains));
+            '') ([ cfg.hostName ] ++ cfg.extraOptions.trusted_domains));
```
Is this still needed btw? I mean the whole `nextcloud-occ call`.

onny replied:
I would like not to touch in this PR before it's getting too complicated and taking too long to review

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -853,11 +852,7 @@ in {
-              ${optionalString (c.overwriteProtocol != null) "'overwriteprotocol' => '${c.overwriteProtocol}',"}
               ${optionalString (c.dbname != null) "'dbname' => '${c.dbname}',"}
               ${optionalString (c.dbhost != null) "'dbhost' => '${c.dbhost}',"}
               ${optionalString (c.dbport != null) "'dbport' => '${toString c.dbport}',"}
```
Can't we switch these things over as well?

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -853,11 +852,7 @@ in {
               ],
               ${optionalString (showAppStoreSetting) "'appstoreenabled' => ${renderedAppStoreSetting},"}
               'datadirectory' => '${datadir}/data',
```
This could be e.g. `services.nextcloud.extraOptions.datadirectory = lib.mkDefault "${datadir}/data";` then.

Ma27 reviewed (commented):
> Merge cfg.hostName into extraOptions.trusted_domains

`services.nextcloud.extraOptions.trusted_domains = [ cfg.hostName ];` ?

Lists are additive.

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -63,27 +63,30 @@ let
-    (mkRemovedOptionModule [ "services" "nextcloud" "disableImagemagick" ] ''
-      Use services.nextcloud.enableImagemagick instead.
-    '')
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "logLevel" ]
+      [ "services" "nextcloud" "extraOptions" "loglevel" ])
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "logType" ]
+      [ "services" "nextcloud" "extraOptions" "log_type" ])
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "defaultPhoneRegion" ]
+      [ "services" "nextcloud" "extraOptions" "default_phone_region" ])
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "overwriteProtocol" ]
+      [ "services" "nextcloud" "extraOptions" "overwriteprotocol" ])
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "skeletonDirectory" ]
+      [ "services" "nextcloud" "extraOptions" "skeletondirectory" ])
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "globalProfiles" ]
+      [ "services" "nextcloud" "extraOptions" "profile.enabled" ])
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "extraTrustedDomains" ]
+      [ "services" "nextcloud" "extraOptions" "trusted_domains" ])
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "trustedProxies" ]
+      [ "services" "nextcloud" "extraOptions" "trusted_proxies" ])
```
```suggestion
    (mkRenamedOptionModule [ "services" "nextcloud" "config" "logLevel" ] [ "services" "nextcloud" "extraOptions" "loglevel" ])
    (mkRenamedOptionModule [ "services" "nextcloud" "config" "logType" ] [ "services" "nextcloud" "extraOptions" "log_type" ])
    (mkRenamedOptionModule [ "services" "nextcloud" "config" "defaultPhoneRegion" ] [ "services" "nextcloud" "extraOptions" "default_phone_region" ])
    (mkRenamedOptionModule [ "services" "nextcloud" "config" "overwriteProtocol" ] [ "services" "nextcloud" "extraOptions" "overwriteprotocol" ])
    (mkRenamedOptionModule [ "services" "nextcloud" "config" "skeletonDirectory" ] [ "services" "nextcloud" "extraOptions" "skeletondirectory" ])
    (mkRenamedOptionModule [ "services" "nextcloud" "config" "globalProfiles" ] [ "services" "nextcloud" "extraOptions" "profile.enabled" ])
    (mkRenamedOptionModule [ "services" "nextcloud" "config" "extraTrustedDomains" ] [ "services" "nextcloud" "extraOptions" "trusted_domains" ])
    (mkRenamedOptionModule [ "services" "nextcloud" "config" "trustedProxies" ] [ "services" "nextcloud" "extraOptions" "trusted_proxies" ])
```

onny commented:
Thank you @SuperSandro2000 and @Ma27 for the review!

onny commented:
Too bad, the changes doesn't seem to work yet. Tests only build after applying this "patch" https://github.com/NixOS/nixpkgs/pull/212430/commits/a7eaf3576cfc964508065ce9fb64093cf6a5c6cd

Ma27 commented:
fyi I may need a few days until I get to review this, so please wait before considering a merge :)

onny commented:
dont know why building nixos manual fails :(

onny commented:
@Ma27 ping :)

AkechiShiro commented:
@onny you could mention your PR here if it's ready for review : https://discourse.nixos.org/t/prs-ready-for-review someone should take it from there

Ma27 commented:
Even though I'm a little busy currently, it'd be cool if I could also have a chance to take a closer look. Will try to do it this week :)

AkechiShiro commented:
@Ma27 I will take a look but this is my first review, I'll need to have a thorough read of review documentation for PRs in NixOS/Nixpkgs manual.

teto commented:
This looks LGTM and https://github.com/NixOS/nixpkgs/pull/245155 makes sense as well. Worth to have in 23.11. I will let the nextcloud maintainer do their stuff but if this helps getting the PR merged, I can try this PR on my nextcloud server.

Ma27 commented:
Agreed, I'll do my best to get this ready for 23.11 :)

Ma27 commented:
I need to set some expectations here: I'll be gone for the next two weeks (and with gone I mean that I plan to not do any Internet stuff in there) and it's becoming quite clear that I won't be able to get to it this week, sorry!

onny committed:
nixos/nextcloud: Move options into now freeForm extraOptions

onny commented:
Rebased on master and put the release notes into 24.05 :)

SuperSandro2000 reviewed (approved):
I don't think I've used any of the options but deploying with it revealed no errors.

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -93,6 +93,22 @@ in {
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "overwriteProtocol" ] [ "services" "nextcloud" "extraOptions" "overwriteprotocol" ])
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "skeletonDirectory" ] [ "services" "nextcloud" "extraOptions" "skeletondirectory" ])
```
```suggestion
      [ "services" "nextcloud" "skeletonDirectory" ] [ "services" "nextcloud" "extraOptions" "skeletondirectory" ])
```

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -93,6 +93,22 @@ in {
       Use services.nextcloud.enableImagemagick instead.
     '')
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "logLevel" ] [ "services" "nextcloud" "extraOptions" "loglevel" ])
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "logType" ] [ "services" "nextcloud" "extraOptions" "log_type" ])
```
```suggestion
      [ "services" "nextcloud" "logLevel" ] [ "services" "nextcloud" "extraOptions" "loglevel" ])
    (mkRenamedOptionModule
      [ "services" "nextcloud" "logType" ] [ "services" "nextcloud" "extraOptions" "log_type" ])
```
those are not under config

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -608,30 +540,111 @@ in {
+            '';
+          };
+          default_phone_region = mkOption {
+            default = "";
```
Why is the default not `null` anymore? If that doesn't work with nextcloud, the type shouldn't be `nullOr` anymore.
Also, does an empty string actually work? I'd prefeer to just use the upstream default here.

onny replied:
Somehow "null" will result in invalid config file, so I removed nullOr

Ma27 replied:
As it doesn't seem to have a default, this seems correct to me now.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -608,30 +540,111 @@ in {
+            '';
+          };
+          profile = {
+            enable = mkEnableOption (lib.mdDoc "global profiles") // {
```
The option must be called `"profile.enabled"`. Nextcloud does not turn dotted option names into nested structures.

Ma27 commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -47,6 +47,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
   The `nimPackages` and `nim2Packages` sets have been removed.
   See https://nixos.org/manual/nixpkgs/unstable#nim for more information.
 
+- Following options of the Nextcloud module are now covered by `extraOptions` and should be migrated: `logLevel` (now `extraOptions.loglevel`), `logType` (`extraOptions.log_type`), `defaultPhoneRegion` (`extraOptions.default_phone_region`), `overwriteProtocol` (`extraOptions.overwriteProtocol`), `skeletonDirectory` (`extraOptions.skeletondirectory`), `globalProfiles` (`extraOptions.profile.enable ...
```
How about "the following options of the Nextcloud module were moved into [`services.nextcloud.extraOptions`](#opt-services.nextcloud.extraOptions) and match the upstream option names"?

Ma27 commented:
Thanks a lot, this is a very nice PR!
Just left a few minor things, but after that being fixed, we should be good to go!

onny commented:
Added the requested changes and tests still suceed 👍

Ma27 committed:
nixos/release-notes: more details about Nextcloud options rename

* Explain why it was renamed (to match the format from `config.php`).
* Show old and new name and link to the option in the options reference.

Ma27 commented:
Pushed a small improvement to the release notes, will merge when CI is green!

emilylange commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -94,6 +94,22 @@ in {
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "skeletonDirectory" ] [ "services" "nextcloud" "extraOptions" "skeletondirectory" ])
+    (mkRenamedOptionModule
+      [ "services" "nextcloud" "config" "globalProfiles" ] [ "services" "nextcloud" "extraOptions" "profile.enabled" ])
```
The option used to be called `services.nextcloud.globalProfiles` not `services.nextcloud.config.globalProfiles`:

```diff
diff --git a/nixos/modules/services/web-apps/nextcloud.nix b/nixos/modules/services/web-apps/nextcloud.nix
index 32de50d08933..cf9560a111fd 100644
--- a/nixos/modules/services/web-apps/nextcloud.nix
+++ b/nixos/modules/services/web-apps/nextcloud.nix
@@ -142,7 +142,7 @@ in {
     (mkRenamedOptionModule
       [ "services" "nextcloud" "skeletonDirectory" ] [ "services" "nextcloud" "extraOptions" "skeletondirectory" ])
     (mkRenamedOptionModule
-      [ "services" "nextcloud" "config" "globalProfiles" ] [ "services" "nextcloud" "extraOptions" "profile.enabled" ])
+      [ "services" "nextcloud" "globalProfiles" ] [ "services" "nextcloud" "extraOptions" "profile.enabled" ])
     (mkRenamedOptionModule
       [ "services" "nextcloud" "config" "extraTrustedDomains" ] [ "services" "nextcloud" "extraOptions" "trusted_domains" ])
     (mkRenamedOptionModule
 ```

Ma27 replied:
Oof good catch!
Do you have some time to file a patch for that?
Otherwise I'll try to do it soonish :)

onny replied:
https://github.com/NixOS/nixpkgs/pull/278478
